File: LeetCode/iter1/c++/633.sum-of-square-numbers.cpp
#include <iostream>
#include <string>
#include <algorithm>
#include <vector>
#include <sstream>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <queue>
#include <stack>

using namespace std;

class Solution {
public:
	bool judgeSquareSum(int c) {
		int upperBound = sqrt(c);
		unordered_set<int> set;
		for (int i = 0; i <= upperBound; i++) {
			int iSquare = i*i;
			set.insert(iSquare);
			if (set.count(c-iSquare) > 0) {
				return true;
			}
		}

		return false;
	}
};

int main()
{
	Solution solution;
	cout << (solution.judgeSquareSum(2) ? "true\n" : "false\n");
	cout << (solution.judgeSquareSum(4) ? "true\n" : "false\n");
	cout << (solution.judgeSquareSum(5) ? "true\n" : "false\n");
	cout << (solution.judgeSquareSum(9) ? "true\n" : "false\n");
	cout << (solution.judgeSquareSum(147) ? "true\n" : "false\n");
	cout << (solution.judgeSquareSum(2147482647) ? "true\n" : "false\n");

	return 0;
}

File: LeetCode/iter1/java/633.sum-of-square-numbers.java
package org.ironman.base.playground.leetcode;

import java.util.HashSet;
import java.util.Set;

public class Solution {

    public static void main(String[] args) {
        System.out.println("0 " + judgeSquareSum(0));
        System.out.println("2 " + judgeSquareSum(2));
        System.out.println("3 " + judgeSquareSum(3));
        System.out.println("4 " + judgeSquareSum(4));
        System.out.println("5 " + judgeSquareSum(5));
        System.out.println("9 " + judgeSquareSum(9));
        System.out.println("147 " + judgeSquareSum(147));
        System.out.println(Integer.MAX_VALUE + " " + judgeSquareSum(Integer.MAX_VALUE));
    }

    public static boolean judgeSquareSum(int c) {
        int upperBound = (int) Math.sqrt(c);
        Set<Integer> squareSet = new HashSet<>();
        for (int i = 0; i <= upperBound; i++) {
            int iSquare = i * i;
            squareSet.add(iSquare);
            if (squareSet.contains(c - iSquare)) {
                return true;
            }
        }

        return false;
    }
}